File: csharp/Tennis/TennisGame3.cs
using System;

namespace Tennis
{
    public class TennisGame3 : ITennisGame
    {
        private int p1Score;
        private int p2Score;
        private readonly string p1Name;
        private readonly string p2Name;
        private static readonly string[] scoreNames = { "Love", "Fifteen", "Thirty", "Forty" };
        private const int EarlyGameScoreLimit = 4;
        private const int MaxPreDeuceScoreSum = 6;

        public TennisGame3(string player1Name, string player2Name)
        {
            this.p1Name = player1Name;
            this.p2Name = player2Name;
        }

        public string GetScore()
        {
            if (IsEarlyGame())
                return EarlyGameScore();
            return EndGameScore();
        }

        private bool IsEarlyGame() => p1Score < EarlyGameScoreLimit && p2Score < EarlyGameScoreLimit && p1Score + p2Score < MaxPreDeuceScoreSum;

        private string EarlyGameScore()
        {
            if (p1Score == p2Score)
                return scoreNames[p1Score] + "-All";
            return scoreNames[p1Score] + "-" + scoreNames[p2Score];
        }

        private string EndGameScore()
        {
            var scoreDifference = Math.Abs(p1Score - p2Score);
            var leaderName = p1Score > p2Score ? p1Name : p2Name;
            return scoreDifference switch
            {
                1 => $"Advantage {leaderName}",
                _ => $"Win for {leaderName}",
            };
        }
        public void WonPoint(string playerName)
        {
            if (playerName == p1Name)
                p1Score++;
            else
                p2Score++;
        }
    }
}

File: java/src/main/java/TennisGame3.java
package main.java;

import main.java.TennisGame;

public class TennisGame3 implements TennisGame {

    // 변수명 알아보기 쉽게 변경
    // 남들이 봤을 때 한 번에 알 수 있어야 함
    private int p2;
    private int p1;
    private String p1N;
    private String p2N;

    public TennisGame3(String p1N, String p2N) {
        this.p1N = p1N;
        this.p2N = p2N;
    }

    public String getScore() {
        String s;

        // 계산식을 함수로 별도로 분리
        // 무슨 계산식인지 다른 사람은 바로 알 수가 없음
        if (p1 < 4 && p2 < 4 && !(p1 + p2 == 6)) {
            String[] p = new String[]{"Love", "Fifteen", "Thirty", "Forty"}; 
            s = p[p1];
            
            // 계산식을 함수로 별도로 분리
            // 무슨 계산식인지 다른 사람은 바로 알 수가 없음
            return (p1 == p2) ? s + "-All" : s + "-" + p[p2];
        } else {
            if (p1 == p2)
                return "Deuce";
            s = p1 > p2 ? p1N : p2N;
            // 계산식을 함수로 별도로 분리
            // 무슨 계산식인지 다른 사람은 바로 알 수가 없음
            return ((p1-p2)*(p1-p2) == 1) ? "Advantage " + s : "Win for " + s;
        }
    }
    
    public void wonPoint(String playerName) {
        // else 없애기
        // 조기 return을 쓰면 else 없어도 됨
        if (playerName == "player1")
            this.p1 += 1;
        else
            this.p2 += 1;
        
    }

}
